data/tribes/scripting/help/ware_help.lua

```diff
@@ -76,44 +76,54 @@ function ware_help_producers_string(tribe, ware_description)
             end
          end
 
-            -- Now collect all wares produced by the filtered programs
-            local produced_wares_strings = {}
-            local produced_wares_counters = {}
-            for j, program_name in ipairs(producing_programs) do
-               local produced_wares_amount = {}
-               produced_wares_counters[program_name] = 0
-               for ware, amount in pairs(building:produced_wares(program_name)) do
-                  if (produced_wares_amount[ware] == nil) then
-                     produced_wares_amount[ware] = 0
-                  end
-                  produced_wares_amount[ware] = produced_wares_amount[ware] + amount
-                  produced_wares_counters[program_name] = produced_wares_counters[program_name] + amount
+         -- Now collect all wares produced by the filtered programs
+         local produced_wares_strings = {}
+         local produced_wares_counters = {}
+         for j, program_name in ipairs(producing_programs) do
+            local produced_wares_amount = {}
+            produced_wares_counters[program_name] = 0
+            for ware, amount in pairs(building:produced_wares(program_name)) do
+               if (produced_wares_amount[ware] == nil) then
+                  produced_wares_amount[ware] = 0
                end
-               local produced_wares_string = ""
-               for ware, amount in pairs(produced_wares_amount) do
+               produced_wares_amount[ware] = produced_wares_amount[ware] + amount
+               produced_wares_counters[program_name] = produced_wares_counters[program_name] + amount
+            end
+            local produced_wares_string = ""
+            for ware, amount in pairs(produced_wares_amount) do
                local ware_descr = wl.Game():get_ware_description(ware)
-                  produced_wares_string = produced_wares_string
-                     .. help_ware_amount_line(ware_descr, amount)
+               produced_wares_string = produced_wares_string
+                  .. help_ware_amount_line(ware_descr, amount)
+            end
+            produced_wares_strings[program_name] = produced_wares_string
+         end
+         -- check for doubled entries
+         for j, program_name in ipairs(producing_programs) do
+            for i, prog_name in ipairs(producing_programs) do
+               if (produced_wares_strings[program_name] == produced_wares_strings[prog_name] and prog_name ~= program_name) then
+                  if (help_consumed_wares_workers(tribe, building, program_name) == help_consumed_wares_workers(tribe, building, prog_name)) then
+                     table.remove(producing_programs, i)
```

Comment:
Removing items from a table we're iterating on is dangerous. It can lead to skipping some items. It happens to work now because we only need at most one removal. If there were more it could break.

Reply:
the idea behind this double loop is to check all programs in the table against the each other starting from the beginning. If any entry in the table is similar to a previous one it does not need to be checked again and can be savely removed this is why we delete the index in the inner loop as it is in advance of the outer loop index. An entry is only deleted if it is a real doubled entry in terms of input and output. Even a changed order of input wares prevents its deletion. So I beleive this is loop is pretty save. As I can't think of any working alternative and I tried a couple of other, any proposal would be appreciated.

Reply:
The inner loop *can* skip entries after a deletion. They *might* be picked up again in a subsequent outer loop iteration, but I'd rather not rely on this...

I would do it like this:
```
         -- check for doubled entries (identical consumed and produced wares)
         local deduplicated_programs = {}
         for j, prog1_name in ipairs(producing_programs) do
            local duplicate = false
            for i, prog2_name in ipairs(deduplicated_programs) do
               if produced_wares_strings[prog1_name] == produced_wares_strings[prog2_name] and
                     help_consumed_wares_workers(tribe, building, prog1_name) ==
                     help_consumed_wares_workers(tribe, building, prog2_name) then
                  duplicate = true
                  break
               end
            end
            if not duplicate then
               table.insert(deduplicated_programs, program_name)
            end
         end
         producing_programs = deduplicated_programs
```

Reply:
ok, this is cleaner. Will change it accordingly. thanks for the tip